### DISCUSSION TITLE
Harmonic cavity

### DISCUSSION THREAD

**Discussion by lnadolski** [atcollab/at#450] [Matlab AT]
Dear @lfarv,

I wonder how to compute the bunch length introduced by a harmonic cavity.
Should I use ohmienvelop to get the new bunch length?
The function atx is doing so, but I got the same results with and without harmonic cavity.

What is the way to do it ?

Best regards,

Laurent.

**Comment by lfarv** [comment]
Hello @lnadolski
"C'est une excellente question, et je vous remercie de l'avoir posée…"
Well, at first glance, I do not see why `ohmienvelope` fails. It should give the bunch length for a linearised RF slope. If you have a simple ring with a harmonic cavity, can you post it?
I would check:
1st, that the `TimeLag` of the harmonic cavity is properly set (~at the synchronous phase),
2nd, that the one turn transfer matrix is indeed affected by the harmonic cavity (5th row).

Otherwise, it should not be too difficult to improve the analytical derivation of bunch length from energy spread in `atsummary`/`ringpara`. At the moment it assumes a single RF system. Instead, getting the linearised RF slope of a complex RF system, knowing the 6D closed orbit, looks easy. This is approximate, but simple. To be done…

Finally, you may consider brute force by using pure tracking. Together with the "Impedance" element (python only, at the moment), this gives results in realistic conditions. Using `fastring` and multiprocessing, it's affordable. I think @swhite2401, @carmignani and @lcarver have experience with that. This reminds me that the documentation for using OpenMP or MPI is still missing. I'll prepare something for that.

**Comment by lnadolski** [comment]
Dear @lfarv,

The issue may come from my side with a wrong assignation of the timelag.
Here a few lines to define and insert the fourth order harmonic cavity:

the harmonic cavity parameters are computing as follows:
index 1: is the main cavity 
index2 : is the HC

```
% first Get main cavity parameters
[circumference, energy, h, iRF, f1, V1] =atGetRingProperties(ring4, 'Circumference', ...
    'Energy', 'harmonic_number', 'cavpts', 'rf_frequency', 'rf_voltage');

% use sine formula to define phases and voltages as in AT

m = 4; m2 = m^2;
V1 = ring4{iRF}.Voltage;
S1 = m2/(m2-1)*atgetU0(ring4)/V1;
phi1 = asin(S1);
phi2 = atan(tan(phi1)/m);
S2 = sin(phi2);
V2 = -V1/m2*S1/S2;
```

the cavity is then defined

```
% Defined fourth harmonic frequency
f2 = f1*4;
elem = atrfcavity('4HC',0,V2,f2,h,energy,'RFCavityPass');
after_ = 1;
ring5 = atinsertelems(ring4, iRF, after_, elem);
```

Time lags are finally defined:
```
bucketdistance = circumference/h;
ring5{iRF+1}.TimeLag = -phi2/(2*pi)*bucketdistance; % HCM
ring5{iRF}.TimeLag = -phi1/(2*pi)*bucketdistance; % main cavity
```

Remarks 
1/ even adding the harmonic m in the time lag does not help
2/ a parametrization search give me maximum bunch lengthening  far from phi2 value.

I would be happy to see what I am doing wrongly.

Other remarks
I notice that the PhaseLag are not yet implemented in the atsetcavity function

**Comment by carmignani** [comment]
Hello @lnadolski,
as @lfarv said, I also think that ohmienvelope should give the solution with the linearised slope. The problem is that if the harmonic cavity is well tuned to give bunch lengthening the linearised slope is zero and I'm not sure that you can trust the results. I tried with a function I  have that puts a harmonic cavity in the lattice and I get with ohmienvelope a bunch length that is approximately 5 times longer with a third harmonic cavity, but again I'm not sure we can trust this result. I think that the linear approximation of the RF slope is not good to get the bunch length in presence of harmonic cavity.
In any case, if ohmienvelope gives exactly the same results with and without harmonic cavity maybe the time lags or the voltages are not correct. Did you consider the fact that the voltages have a minus sign in the AT cavity convention?
As @lfarv said, you can get easily the bunch lengthening by performing multi particle tracking with the quantum fluctuation and with the fast ring, even on a single core.

**Comment by lfarv** [comment]
Hi @lnadolski
I did'n fully understand your tuning, but here are a few comments:

1. The simplest way to set the `TimeLag` of the harmonic cavities is to set `Timelag = closed_orbit(6);`. Looking at the longitudinal kick in `RFCavityPass.c`:

   ```C
   r6[4] += -nv*sin(TWOPI*freq*((r6[5]-lag)/C0 -...
   ```
   this ensures that the zero crossing of the harmonic cavity will be at the beam location. It's easy and safe.
   Concerning the main cavity, my preference is to keep `TimeLag=0` , so that the main cavity stays as the time reference, which is the definition of AT coordinate system.
   Otherwise, some people prefer detuning the main cavity so that the stable longitudinal time is zero (`atSetCavityPhase` does that). Then the `TimeLag` of harmonic cavities is zero (and `Timelag = closed_orbit(6);` still works). This is a change of referential, with possible unwanted side effects.
   Dephasing both main and harmonic cavities looks to me an unnecessary complication!
2. `PhaseLag` was introduced in #348 to have an input for beam loading control. I discourage using it for other purposes: this would defeat the original intention. Anyway, as described above, controlling in time is more convenient than in phase.
3. I fully agree with @carmignani's remarks about linearisation: for a zero slope, this may overestimate the result. For a partial compensation, maybe this is less relevant.

**Comment by carmignani** [comment]
Hello @lfarf,
as I wrote before, you have to be careful with the sign. Whatever you do with the TimeLag of the main cavity, the the TimeLag of the harmonic cavity can be closed_orbit(6) only if you give a negative voltage to the harmonic cavity, otherwise it must be closed_orbit(6) + half a harmonic cavity period. 
This setting of the TimeLags gives a zero first derivative of the voltage at the synchronous phase, but a negative second derivative. Some text books/papers suggest to have also a zero second derivative and in order to do that you have to change the phases a little bit.

**Reply by lfarv** [reply]
Right. You could even say that more generally, to have the zero crossing at the beam location, you need
TimeLag=closed_orbit(6) + k (c.T<sub>h</sub>/2)
with any integer k. Even k needs on sign for V, odd k needs the other sign.

In my opinion, taking k=0 is the simplest solution, you just choose the sign of V which reduces the slope.

**Comment by lnadolski** [comment]
Dear @lfarv 

Thank you for your feedback.
As stated by @carmignani, they are three conditions to be met for setting the harmonic cavity

- the voltage at the synchronous phase should be equal to the energy lost per turn
- the first and second derivatives of the voltage should be zero  at the synchronous phase

This results with a condition on the RF voltage of the harmonic cavity and set the values of both the main RF system and the harmonic RF system. See for example [this reference](https://accelconf.web.cern.ch/icap2018/papers/tupaf12.pdf) by G. Bassi

 Results given by atx and tracking are indeed quite different.

**Reply by lnadolski** [reply]
@lfarv 
I have noticed that findorbit6 (xorbit°6) is written assuming that the time_lag is 0. This may result in not finding the closed orbit.

**Reply by swhite2401** [reply]
This used to be an issue but has be fixed. Please have a look at `physics.orbit._orbit6`, if no initial guess is provided the timelag is computed using tracking

**Reply by lfarv** [reply]
@lnadolski : yes, you are right, the initial guess assumes `TimeLag=0`. Another inconvenience of setting the `TimeLag` of main cavities different from 0 ! Anyway, this can be easily corrected, I put it on the list !

**Reply by swhite2401** [reply]
@lfarv  `get_timelag_from_U0` does not make any assumption on the initial `TimeLag` right?

**Reply by lfarv** [reply]
@swhite2401: yes,sorry I did not make it clear, in python it's correct, we were talking of `xorbit_6` in Matlab !

**Reply by swhite2401** [reply]
Ah my bad, sorry for the confusion then!

**Comment by lfarv** [comment]
> As stated by @carmignani, they are three conditions to be met for setting the harmonic cavity
>  
> - the voltage at the synchronous phase should be equal to the energy lost per turn
> - the first and second derivatives of the voltage should be zero at the synchronous phase
This results with a condition on the RF voltage of the harmonic cavity and set the values of both the main RF system and the harmonic RF system. See for example [this reference](https://accelconf.web.cern.ch/icap2018/papers/tupaf12.pdf) by G. Bassi

I fully agree, but you are still free to shift the `TimeLag` of **all** cavities by the same arbitrary amount. This is where you have two simple solutions:

- either freeze the `TimeLag` of the main cavities to zero: **the main cavities are the time reference**, all phases, timelags, 6<sup>th</sup> particle coordinate,… are measured with respect to the main cavities. x<sub>co</sub>(6) is the synchronous phase, and the `TimeLag` of harmonic cavities is then equal to x<sub>co</sub>(6), or close to that, to follow @carmignani. This looks like to real life, where the RF clock is usually the trigger for time measurements,
- or set the reference at the synchronous phase, so that **the closed orbit is the reference** (x<sub>co</sub>(6) == 0). Then all phases, timelags,... are measured with respect to the beam. In that case the `TimeLag` of the harmonic cavities is zero, or close to zero. The `TimeLag` of the main cavities is the opposite of the synchronous phase. But keep in mind that this is a moving referential, depending on the RF voltage, on beam current if you have longitudinal impedances, etc. That's why I prefer the 1<sup>st</sup> solution, but it's a matter of taste !

Any other choice makes life more complicated !
